File: GFG/BST/mergeBSTs.c
// import java.io.*;
// import java.lang.*;
// import java.util.*;

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include<stdbool.h>
/*

# Author			: @RAJ009F
# Topic or Type 	: GFG/BST
# Problem Statement	: Java: Merge two BSTs 
# Description		: 
# Complexity		: 
=======================
#sample output
----------------------

=======================
*/
struct BT
{
	int data;
	struct BT *l;
	struct BT *r;	
};

struct Stack{
	
	struct BT *data;
	struct Stack *next;
	
};

struct Stack *createS()
{
	struct Stack *s = (struct Stack *)malloc(sizeof(struct Stack));
	if(!s)
		return;
	s->data = NULL;
	s->next = NULL;
	
	return s;
}

struct Stack *newNode(struct BT *data)
{
	struct Stack *node =  (struct Stack *) malloc(sizeof(struct Stack));
	if(!node)
		return NULL;
	node->data = data;
	node->next = NULL;
	
	return node;
}


void push(struct Stack **s,  struct BT *data)
{
	struct Stack *temp = newNode(data);
	if(!(*s))
	{
		
		
		*s = temp;
	}else
	{
		temp->next = *s;
		*s = temp;
	}
	
	
}

bool isEmpty(struct Stack *s)
{
	return s == NULL;
}

struct BT  *pop(struct Stack **s)
{
	if(isEmpty(*s))
		return NULL;
	struct BT *data = (*s)->data;
	struct Stack *t = *s;
	*s = (*s)->next;
	
	free(t);
	return data;	
}


void inorder(struct BT *root)
{
	printf("\ninorder called");
	if(root)
	{
		inorder(root->l);
		printf("%d\t", root->data);
		inorder(root->r);
		
	}
	
	
}

void mergeTrees(struct BT *root1, struct BT *root2)
{
	if(!root1 && !root2)
		return;
	
	if(!root1)
	{
		inorder(root2);
		return;	
	}
	
	if(!root2)
	{
		inorder(root1);
		return;		
	}
	
	struct Stack *s1 = createS();
	struct Stack *s2 = createS();
	
	while(!isEmpty(s1) || !isEmpty(s2) || root1 || root2)
	{
		
		if(root1 || root2)
		{
			if(root1)
			{
				push(&s1, root1);
				root1 = root1->l;
				
			}
			if(root2)
			{
				push(&s2, root2);
				if(root2->data == 6)
					printf("6");
				root2 = root2->l;
				
			}
			
			
			
		}else
		{
			if(isEmpty(s1))
			{
				printf("kkk");
				while(!isEmpty(s2));
				{
					root2 = pop(&s2);
					root2->l = NULL;
					inorder(root2);
				}
				return;
			}
			
			if(isEmpty(s2))
			{
				while(!isEmpty(s1));
				{
					root1 = pop(&s1);
					root1->l = NULL;
					inorder(root1);
				}
				return;
			}
			
			
			root1  =  pop(&s1);
			root2 = pop(&s2);
			
			if(root2->data > root1->data)
			{
				printf("%d ", root1->data);
				root1 = root1->r;
				//printf("sdsd");
				push(&s2, root2);
				//printf("sdsd");
				root2 = NULL;
				
			}else
			{
				printf("%d ", root2->data);
				root2 = root2->r;
				push(&s1, root1);
				root1 = NULL;
				
				
			}
				
			
			
			
		}
		
		
		
		
		
	}
	
	
	
	
}

struct BT *newNodeT(int data)
{
	struct BT *node  =  (struct BT *) malloc(sizeof(struct BT));
	node -> data = data;
	node -> l = node->r = NULL;
	
	return node;
	
	
}

int main()

{
	struct BT *root1 = NULL;
	struct BT *root2 = NULL;
	
	root1 = newNodeT(3);
	root1->l = newNodeT(1);
	root1->r = newNodeT(5);
	
	
	root2 = newNodeT(4);
	root2->l = newNodeT(2);
	root2->r = newNodeT(6);
	
	mergeTrees(root1, root2);
	
	return 0;
}

File: GFG/BST/MergeBSTs.java
import java.io.*;
import java.lang.*;
import java.util.*;

// #include <stdio.h>
// #include <stdlib.h>
// #include <string.h>
// #include<stdbool.h>
/*

# Author			: @RAJ009F
# Topic or Type 	: GFG/BST
# Problem Statement	: Java: Merge two BSTs 
# Description		: 
# Complexity		: 
=======================
#sample output
----------------------

=======================
*/

class Node {
	
	int data;
	Node l;
	Node r;
	
	Node(int data)
	{
		
		this.data = data;
		this.l = this.r = null;
		
	}

}


class MergeBSTs

{

static Node root ;


public void inorder(Node root)
{
	if(root != null)
	{
		inorder(root.l);
		System.out.print(root.data + " ");
		inorder(root.r);
		
	}
	
	
}


public void mergeTrees(Node root1, Node root2)
{
	if(root1 == null && root2 == null)
		return;
	if(root1 == null)
	{
		inorder(root2);
		return;
	}
	if(root2 == null)
	{
		inorder(root1);
		return;
	}
	
	Stack<Node> s1 =  new Stack<Node>();
	Stack<Node> s2 = new Stack<Node>();
	
	while(root1 !=null || root2 != null || s1.isEmpty() == false || s2.isEmpty()==false )
	{
		if( root1 == null || root2 == null)
		{
		if(root1 != null)
		{
			s1.push(root1);
			root1 = root1.l;
			
			
		}
		
		if(root2 != null)
		{
			s2.push(root2);
			root2 = root2.l;
		}
		
		}
		else{		
			if(root1 == null)
			{				
		     while(s2.isEmpty() == false)
				{
					root2 = s2.pop();
					root2.l = null;
					inorder(root2);				
				}				
				return;
			}

			if( root2 == null)
			{
				while(s1.isEmpty() == false)
				{
					root1 = s1.pop();
					root1.l = null;
					inorder(root1);
					
				}				
				return;
				
			}
		
			
			root1  =  s1.pop();
			root2  =   s2.pop();
			
			if(root1.data> root2.data)
			{
				System.out.print(root2.data+" ");
				root2 = root2.r;
				s1.push(root1);
				root1 = null;
			}else
			{
				System.out.print(root1.data+" ");
				root1 = root1.r;
				s2.push(root2);
				root2 = null;
				
			}
			
			
		
		
		
		}
		
		
		
		
		
		
		
		
	}
	
	
}

public void mergeTrees1(Node root1, Node root2)
{
	if(root1 == null && root2 == null)
			return;
	
	if(root1 == null)
	{
		inorder(root2);
		return;
	}
	
	if(root2 == null)
	{
		inorder(root1);
		return;	
	}
	
	Stack<Node> s1 =  new Stack<Node>();
	Stack<Node> s2 = new Stack<Node>();
	
	boolean flag1= true;
	boolean flag2 = true;
	
	while( flag1 || flag2 || s1.isEmpty() == false || s2.isEmpty()==false)
	{
		while(flag1)
		{
			if(root1 != null)
			{
				s1.push(root1);
				root1 =  root1.l;		
			}else
			{
				if(s1.isEmpty())
					flag1 = false;
				else
				{
					root1 = s1.pop();
					flag1 = false;
				}	
			}			
		}
		
		while(flag2)
		{
			if(root2 != null)
			{
				s1.push(root2);
				root2 =  root2.l;		
			}else
			{
				if(s2.isEmpty())
					flag2 = false;
				else
				{
					root2 = s2.pop();
					flag2 = false;
				}	
			}			
		}
		if(root1 == null || root2 == null)
		{		
			if(root1 == null && s1.isEmpty())
			{
				
				s2.push(root2);
				while(s2.isEmpty() == false)
				{
					root2 = s2.pop();
					root2.l = null;
					inorder(root2);
					
				}
				return ;
			}
			
			if(root2 == null && s2.isEmpty())
				{
					
					s1.push(root1);
					while(s1.isEmpty() == false)
					{
						root1 = s1.pop();
						root1.l = null;
						inorder(root1);
						
					}
					return ;
				}
		}else
		{
			if(root1.data == root2.data)
			{
				System.out.print(root1.data +" "+root2.data);
				root1 = root1.r;
				root2 = root2.r;
				flag1 = true;
				flag2 = true;
			}else if(root1.data > root2.data)
			{
				System.out.print(root2.data+" ");
				flag2 = true;
				root2 = root2.r;
			}else
			{
				System.out.print(root1.data+" ");
				flag1 = true;
				root1 = root1.r;
			}
			
			
			
			
		}
		
	}
	
	
	
}

public static void main(String args[])
{
	MergeBSTs tree =  new MergeBSTs();
	
	Node root1 = new Node(10);
	root1.l = new Node(4);
	root1.r = new Node(12);
		
	Node root2 =  new Node(32);
	root2.l = new Node(3);
	root2.r =  new Node(45);
	
	tree.mergeTrees(root1, root2);
	
	
}

}